Add HasCallStack constraints to get informative errors and traces.

facundominguez opened the pull request:
This improves the information in error messages, and would be necessary for tracing & analysis in jni.

Also has a commit to check for null pointers which could crash the program. Placed here to avoid conflicts but could be moved to another PR.

facundominguez commented:
There is a penalty in performance.

Adding `HasCallStack` interferes with caching class references. I modified the benchmarks so they allocate the class references in a CAF and pass them to the `call` and `callStatic`.

With HasCallStack constraints:
```
benchmarking Calls/Java calls/static method call: unboxed single arg / unboxed return
time                 608.4 ns   (606.7 ns .. 610.6 ns)

benchmarking Calls/Java calls/jni static method call: unboxed single arg / unboxed return
time                 572.6 ns   (570.5 ns .. 575.1 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)

benchmarking Calls/Java calls/method call: no args / unboxed return
time                 6.003 μs   (5.876 μs .. 6.119 μs)

benchmarking Calls/Java calls/method call: boxed single arg / unboxed return
time                 13.89 μs   (13.15 μs .. 14.89 μs)
```

Without HasCallStack (i.e. without this PR), these are the numbers:
```
benchmarking Calls/Java calls/static method call: unboxed single arg / unboxed return
time                 521.9 ns   (519.9 ns .. 523.9 ns)
             
benchmarking Calls/Java calls/jni static method call: unboxed single arg / unboxed return
time                 499.8 ns   (498.2 ns .. 501.4 ns)
             
benchmarking Calls/Java calls/method call: no args / unboxed return
time                 1.711 μs   (1.701 μs .. 1.722 μs)
             
benchmarking Calls/Java calls/method call: boxed single arg / unboxed return
time                 7.159 μs   (6.367 μs .. 8.197 μs)
```

facundominguez commented:
I suppose most of the cost should be explained by an increment of the calls to the GC, and I'll try to check it.

The alternatives I see:
1. Live with the performance penalty.
2. Ask GHC devs if it would be possible to disable `HasCallStack` with runtime flags.
3. Implement a library to collect stack traces with a mechanism that can be disabled at runtime.
4. Don't collect stack traces, and only collect the source location of call sites.

mboes commented:
Insofar as we can live with (4), that would be ideal. Collecting call stacks is brittle to begin with, because it infects user code too, not just JNI bindings, `jvm` functions, inline-java, sparkle... etc etc. I believe that there are by now alternative ways to get stack traces when we do need them, in particular as a byproduct of the recent DWARF support effort.

Let's ping @bgamari and @scpmw about that.

mboes commented:
> Insofar as we can live with (4), that would be ideal.

Well I should say... least bad, in my view.

bgamari commented:
Right, since 8.2 you can reliably extract backtraces from running Haskell programs compiled with DWARF information (i.e. with GHC's `-g` flag). This functionality is exposed in Haskell through the `GHC.ExecutionStack` [module](https://hackage.haskell.org/package/base-4.10.0.0/docs/GHC-ExecutionStack.html). Note that there are a few caveats here,
 * GHC must be built with `libdw` support enabled (i.e. with the `--enable-dwarf-unwind` `configure` flag)
 * For reliable backtraces, all libraries linked against (including core libraries like `base`) must be built with `-g`
 * These are DWARF backtraces; the source spans will generally be more useful than the symbol names since the latter will often be compiled-generated unique names.

mboes commented:
@bgamari many thanks for you input. A few followup questions:

* Is `libdw` support enabled by default?
* Does this work well even with pervasive use of FFI calls?
* Will DWARF backtraces contain only symbols names, and if so is there an easy way to map those back to source spans?

bgamari commented:
> Is libdw support enabled by default?

We don't enable GHC's unwinding support by default since it requires that `libdw` be installed on the host.  I did, however, produce an [8.2.1 binary distribution](https://downloads.haskell.org/~ghc/8.2.1/ghc-8.2.1-x86_64-deb8-linux-dwarf.tar.xz) with it enabled (and all libraries built with `-g`).

> Does this work well even with pervasive use of FFI calls?

Yes, this was broken in 8.0 (which is why I advised against using it) but 8.2 should be solid.

> Will DWARF backtraces contain only symbols names, and if so is there an easy way to map those back to source spans?

The Haskell interface provides both symbol names and spans. See the haddocks for `GHC.ExecutionStack`.

mboes commented:
Awesome!

Slightly off-topic question: @bgamari I gather from what you're saying that, given the quality of DWARF support now, on a host with libdw etc, in principle we could have a version of `base` where `error` provides good stack traces yet without needing `HasCallStack` constraints that pollute the whole call chain?

bgamari commented:
Yep, this should indeed be possible. In the future it might also be possible to collect stack traces for arbitrary exceptions as well. I jotted down some ideas on this subject on the [Wiki](https://ghc.haskell.org/trac/ghc/wiki/Exceptions/StackTraces).

facundominguez commented:
I tried the ghc binary with dwarf support. Doesn't seem to provide enough information yet.
```
$ cat t.hs
import GHC.ExecutionStack

f = getStackTrace

g = f

main = g >>= putStrLn . maybe "No stack trace" showST

showST = unlines . map (maybe "no loc info" showSrcLoc . srcLoc)

showSrcLoc (SrcLoc f l c) = concat [f, ":", show l, ":", show c]

$ ghc -g -rtsopts -fforce-recomp t.hs
[1 of 1] Compiling Main             ( t.hs, t.o )
Linking t ...

$ ./t
t: Failed to get stack frames of current process: no matching address range: Success
rts/Libdw.c:288:0
no loc info
no loc info
no loc info
no loc info
rts/Libdw.c:259:0
no loc info
```

facundominguez commented:
Rebased master and added a patch with the changes I did to benchmark this.

The GC does have more work to do with `HasCallStack`.
```
$ stack --nix build jvm --bench --benchmark-arguments='-m glob "**/[sm]*ethod call*/*"'
...
  21,658,797,984 bytes allocated in the heap
     868,177,872 bytes copied during GC
       1,940,968 bytes maximum residency (460 sample(s))
         547,936 bytes maximum slop
               6 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     28316 colls,     0 par    0.921s   0.918s     0.0000s    0.0005s
  Gen  1       460 colls,     0 par    0.468s   0.468s     0.0010s    0.0047s

  TASKS: 5 (1 bound, 4 peak workers (4 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time   17.712s  ( 16.123s elapsed)
  GC      time    1.388s  (  1.386s elapsed)
  EXIT    time    0.001s  (  0.000s elapsed)
  Total   time   19.195s  ( 17.509s elapsed)

  Alloc rate    1,222,839,352 bytes per MUT second

  Productivity  92.8% of total user, 92.1% of total elapsed
```
without `HasCallStack`:
```
$ stack --nix build jvm --bench --benchmark-arguments='-m glob "**/[sm]*ethod call*/*"'
...
  12,803,902,856 bytes allocated in the heap
      34,427,952 bytes copied during GC
       2,066,576 bytes maximum residency (14 sample(s))
         621,040 bytes maximum slop
               6 MB total memory in use (0 MB lost due to fragmentation)

                                     Tot time (elapsed)  Avg pause  Max pause
  Gen  0     18123 colls,     0 par    0.133s   0.129s     0.0000s    0.0003s
  Gen  1        14 colls,     0 par    0.022s   0.022s     0.0016s    0.0037s

  TASKS: 5 (1 bound, 4 peak workers (4 total), using -N1)

  SPARKS: 0 (0 converted, 0 overflowed, 0 dud, 0 GC'd, 0 fizzled)

  INIT    time    0.001s  (  0.001s elapsed)
  MUT     time   17.682s  ( 17.429s elapsed)
  GC      time    0.155s  (  0.151s elapsed)
  EXIT    time    0.001s  (  0.001s elapsed)
  Total   time   17.969s  ( 17.582s elapsed)

  Alloc rate    724,114,250 bytes per MUT second

  Productivity  99.1% of total user, 99.1% of total elapsed
```

bgamari commented:
On September 8, 2017 10:12:19 PM GMT+01:00, "Facundo Domínguez" <notifications@github.com> wrote:
>I tried the ghc binary with dwarf support. Doesn't seem to provide
>enough information yet.
>```
>$ cat t.hs
>import GHC.ExecutionStack
>
>f = getStackTrace
>
>g = f
>
>main = g >>= putStrLn . maybe "No stack trace" showST
>
>showST = unlines . map (maybe "no loc info" showSrcLoc . srcLoc)
>
>showSrcLoc (SrcLoc f l c) = concat [f, ":", show l, ":", show c]
>
>$ ghc -g -rtsopts -fforce-recomp t.hs
>[1 of 1] Compiling Main             ( t.hs, t.o )
>Linking t ...
>
>$ ./t
>t: Failed to get stack frames of current process: no matching address
>range: Success
>rts/Libdw.c:288:0
>no loc info
>no loc info
>no loc info
>no loc info
>rts/Libdw.c:259:0
>no loc info
>```

Hmm, interesting. I'll try to reproduce this when I get a chance.
